### PR TITLE
FIX: Address Mac CGO Crash

### LINKED ISSUE #3550: macOS panic when serving middleware response
### Description

I have an in-development Wails application that at one point, makes an HTTP request to a custom path. In my case the request comes from the `src` attribute of an `img` tag set using JavaScript, but I don't think that's relevant. The request is served by my custom code, set in the app options with `assetserver.Options.Middleware`.

Here is the stack trace I get from macOS (formatting is slightly garbled, sorry):

```
panic: runtime error: cgo argument has Go pointer to unpinned Go pointer

goroutine 29 [running]:github.com/wailsapp/wails/v2/pkg/assetserver/webview.(*responseWriter).Write.func1(0xc0000160a0?, 0xc000170548, 0x8)
	/Users/runner/go/pkg/mod/github.com/wailsapp/wails/v2@v2.8.2/pkg/assetserver/webview/responsewriter_darwin.go:108 +0x4bgithub.com/wailsapp/wails/v2/pkg/assetserver/webview.(*responseWriter).Write(0xc000012018, {0xc000170548, 0x8, 0x7fd374b25350?})
	/Users/runner/go/pkg/mod/github.com/wailsapp/wails/v2@v2.8.2/pkg/assetserver/webview/responsewriter_darwin.go:108 +0x6agithub.com/wailsapp/wails/v2/pkg/assetserver.(*contentTypeSniffer).Write(0xc000012030, {0xc000170548, 0x8, 0x8})
	/Users/runner/go/pkg/mod/github.com/wailsapp/wails/v2@v2.8.2/pkg/assetserver/content_type_sniffer.go:19 +0x48
image/png.(*encoder).writeChunk(0xc000170508, {0xc000170554, 0xc000012030?, 0x400}, {0x103e64b08?, 0x8?})
	/Users/runner/hostedtoolcache/go/1.22.4/arm64/src/image/png/writer.go:115 +0x13f
image/png.(*encoder).writeIHDR(0xc000170508)
	/Users/runner/hostedtoolcache/go/1.22.4/arm64/src/image/png/writer.go:166 +0xd8
image/png.(*Encoder).Encode(0x1044945c0, {0x1048a9318, 0xc000012030}, {0x1042704b0, 0xc0003ae080})
	/Users/runner/hostedtoolcache/go/1.22.4/arm64/src/image/png/writer.go:659 +0x539
main.(*App).middleware-fm.(*App).middleware.func1({0x104270ad0?, 0xc000012030?}, 0x7fd374b25300?)
	/Users/runner/work/<censored>/<censored>/app.go:70 +0xcb
net/http.HandlerFunc.ServeHTTP(0xc00017e240?, {0x104270ad0?, 0xc000012030?}, 0xf?)
	/Users/runner/hostedtoolcache/go/1.22.4/arm64/src/net/http/server.go:2166 +0x29github.com/wailsapp/wails/v2/pkg/assetserver.(*AssetServer).ServeHTTP(0xc000122c80, {0x104270ad0, 0xc000012030}, 0xc0003a2120)
	/Users/runner/go/pkg/mod/github.com/wailsapp/wails/v2@v2.8.2/pkg/assetserver/assetserver.go:183 +0x579github.com/wailsapp/wails/v2/pkg/assetserver.(*AssetServer).processWebViewRequestInternal(0xc000122c80, {0x1042724f8, 0xc000126558})
	/Users/runner/go/pkg/mod/github.com/wailsapp/wails/v2@v2.8.2/pkg/assetserver/assetserver_webview.go:149 +0xab7github.com/wailsapp/wails/v2/pkg/assetserver.(*AssetServer).processWebViewRequest(0xc000122c80, {0x1042724f8, 0xc000126558})
	/Users/runner/go/pkg/mod/github.com/wailsapp/wails/v2@v2.8.2/pkg/assetserver/assetserver_webview.go:57 +0x48
created by github.com/wailsapp/wails/v2/pkg/assetserver.(*AssetServer).ServeWebViewRequest in goroutine 19
	/Users/runner/go/pkg/mod/github.com/wailsapp/wails/v2@v2.8.2/pkg/assetserver/assetserver_webview.go:49 +0xe5
```

(This is a stack trace from v2.8.2 but from looking at the source code I *believe* this still happens with v2.9.1)

The actual panic appears to be happening right here: https://github.com/wailsapp/wails/blob/5cd0caccc4e238e276611e7ed7fc91eac4765c46/v2/pkg/assetserver/webview/responsewriter_darwin.go#L101-L108

I did some research, and the bug here appears to simply be that `content` is a variable, rather than passing the `unsafe.Pointer(&buf[0])` code into the C function directly. Ordinarily this wouldn't matter of course, but it seems for CGo it does because the Go type checker is not smart enough to understand this.

My sources for this claim are:
- https://github.com/golang/go/issues/14210#issuecomment-346402945
- https://github.com/golang/go/issues/28606
- https://ericchiang.github.io/post/cgo/

This would make the fix pretty simple: just get rid of the `content` variable. And change where the same thing happens again [here](https://github.com/wailsapp/wails/blob/5cd0caccc4e238e276611e7ed7fc91eac4765c46/v2/pkg/assetserver/webview/responsewriter_darwin.go#L126-L133). Maybe it happens in other places too?

I have not been able to test this proposed fix yet, or created a minimal reproducible example yet. I hope to do, maybe this weekend. I am filing this issue in advance in hopes that someone else will beat me to it.

I searched the Internet as best I could for anyone running into this specific panic with Wails, and I couldn't find anything. I guess no one has used middleware on macOS with Wails v2 before? Seems possible, but maybe something deeper is going on as well. Hopefully someone else can reproduce this. For what it's worth I have seen this on two separate macOS Sonoma machines.

### To Reproduce

1. Create a wails project with a middleware server, and a frontend that makes a request to that middleware
2. Observe the crash on macOS

Minimal reproducible example repo coming soon hopefully.

### Expected behaviour

No crashing with middleware usage on macOS :)

### Screenshots

Here's basically the same stack trace, but in screenshot form:

image

### Attempted Fixes

Will be attempting a fix with modified C code soon. Essentially just inlining the `content` variable.

### System Details

image

### Additional context

_No response_

### PR DESCRIPTION
**Pull request by leaanthony**
# Description

This fix copies Go memory to C memory to ensure that it does not move during a request.

Fixes #3550 

## Summary by CodeRabbit

- **Bug Fixes**
  - Resolved a CGO memory issue on Darwin, ensuring more stable and efficient memory management.

### CODE REVIEW

**Comment by leaanthony** [comment]
Please try the above PR to see if this fixes your problem. If you are unsure how to do this, please follow [this guide](https://wails.io/docs/guides/local-development#testing-a-pr). Let us know how you get on 👍

File: v2/pkg/assetserver/webview/responsewriter_darwin.go [hunk 1]
```diff
@@ -98,16 +98,28 @@ func (rw *responseWriter) Write(buf []byte) (int, error) {
...
-		content = unsafe.Pointer(&buf[0])
 		contentLen = len(buf)
 	}
 
-	if !C.URLSchemeTaskDidReceiveData(rw.r.task, content, C.int(contentLen)) {
-		return 0, errRequestStopped
+	if contentLen > 0 {
+		// Create a C array to hold the data
+		cBuf := C.malloc(C.size_t(contentLen))
+		defer C.free(cBuf)
```
[the remark this answers is not in the document]
Comment (leaanthony) [reply v2/pkg/assetserver/webview/responsewriter_darwin.go:109]: Return an explicit error created by fmt.Errorf.

[the remark this answers is not in the document]
Comment (leaanthony) [reply v2/pkg/assetserver/webview/responsewriter_darwin.go:109]: Make it a commitable comment

File: v2/pkg/assetserver/webview/responsewriter_darwin.go [hunk 2]
```diff
@@ -98,16 +98,31 @@ func (rw *responseWriter) Write(buf []byte) (int, error) {
...
 		contentLen = len(buf)
 	}
 
-	if !C.URLSchemeTaskDidReceiveData(rw.r.task, content, C.int(contentLen)) {
-		return 0, errRequestStopped
+	if contentLen > 0 {
+		// Create a C array to hold the data
+		cBuf := C.malloc(C.size_t(contentLen))
+		if cBuf == nil {
+		    return 0, fmt.Errorf("memory allocation failed for %d bytes", contentLen)
```
Comment (makew0rld) [inline v2/pkg/assetserver/webview/responsewriter_darwin.go:110]: `fmt` import is missing so compile fails. Fixed this in my fork: https://github.com/wailsapp/wails/compare/v2-bugfix/3550-macos-cgo-error...makew0rld:wails:patch-1

**Review by makew0rld — APPROVED** [review]

**Comment by makew0rld** [comment]
@leaanthony not sure what's up with auto-merge but this is good to be merged. Would appreciate a release but no problem if it's not convenient.

**Comment by leaanthony** [comment]
Thanks for chasing this up!

**Pull request merged**